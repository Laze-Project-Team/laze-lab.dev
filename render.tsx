import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
