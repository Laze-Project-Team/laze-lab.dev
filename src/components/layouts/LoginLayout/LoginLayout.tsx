import { css } from '@emotion/react';
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import Container from '@mui/material/Container';
import { useTranslation } from 'next-i18next';
import type { FC, ReactNode } from 'react';

import { DefaultLink } from '@/components/ui/DefaultLink';
import { pagesPath } from '@/lib/$path';

export type loginLayoutProps = {
  title: ReactNode;
  children?: ReactNode;
};

export type presentialLoginLayoutProps = loginLayoutProps;

export const PresentialLoginLayout: FC<presentialLoginLayoutProps> = ({
  title,
  children,
}) => {
  const [t] = useTranslation('common');

  return (
    <Container maxWidth="sm">
      <main>
        <h1
          css={css`
            margin: 2rem 0;
            font-size: 1.5rem;
            text-align: center;
          `}
        >
          {title}
        </h1>

        <DefaultLink href={pagesPath.$url()}>
          <ArrowBackIosIcon sx={{ fontSize: '1rem' }} />
          <span>{t('backToHome')}</span>
        </DefaultLink>

        {children}
      </main>
    </Container>
  );
};

export const LoginLayout: FC<loginLayoutProps> = (props) => {
  return <PresentialLoginLayout {...props} />;
};
