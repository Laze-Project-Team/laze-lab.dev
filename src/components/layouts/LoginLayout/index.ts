export type { loginLayoutProps } from './LoginLayout';
export { LoginLayout } from './LoginLayout';
