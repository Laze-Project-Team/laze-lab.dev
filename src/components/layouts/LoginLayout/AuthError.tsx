import { Alert } from '@mantine/core';
import type { FC, ReactNode } from 'react';
import { createContext, useState } from 'react';

import { useAuthError } from '@/components/layouts/LoginLayout/useAuthError';

export type authErrorContextType = {
  error: string | null;
  setError: (error: string | null) => void;
};

export const authErrorContext = createContext<authErrorContextType | null>(
  null,
);

export type authError = {
  code: string;
  message: string;
};

export type authErrorsProviderProps = {
  children: ReactNode;
};

export const AuthErrorProvider: FC<authErrorsProviderProps> = ({
  children,
}) => {
  const [error, setError] = useState<string | null>(null);

  return (
    <authErrorContext.Provider value={{ error, setError }}>
      {children}
    </authErrorContext.Provider>
  );
};

export const AuthError: FC = () => {
  const { error, setError } = useAuthError();

  return <>{error && <Alert onClose={() => setError(null)}>{error}</Alert>}</>;
};
