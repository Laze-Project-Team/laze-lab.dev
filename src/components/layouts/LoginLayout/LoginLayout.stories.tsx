import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { PresentialLoginLayout } from './LoginLayout';

export default {
  title: 'layouts/LoginLayout',
  component: PresentialLoginLayout,
} as ComponentMeta<typeof PresentialLoginLayout>;

const Template: ComponentStory<typeof PresentialLoginLayout> = (props) => (
  <PresentialLoginLayout {...props} />
);

export const Primary = Template.bind({});
Primary.args = {
  title: 'Here is the title',
  children: 'Here is the children',
};
