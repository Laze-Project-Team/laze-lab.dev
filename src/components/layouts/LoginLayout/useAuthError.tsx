import { useContext } from 'react';

import type { authError } from './AuthError';
import { authErrorContext } from './AuthError';

export type authErrorHandlers = {
  error: string | null;
  setError: (errorMessage: string | null) => void;
  handleError: (error: authError) => void;
};

export const useAuthError = (): authErrorHandlers => {
  const authCtx = useContext(authErrorContext);

  if (authCtx === null) {
    throw new Error('Please wrap the component with AuthErrorProvider');
  }

  const { error, setError } = authCtx;

  return {
    error,
    setError: (errorMessage) => {
      setError(errorMessage);
    },
    handleError: (err) => {
      setError(err.message);
    },
  };
};
