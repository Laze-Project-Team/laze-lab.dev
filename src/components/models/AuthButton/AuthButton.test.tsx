import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import type { authMethod, authType } from '@/components/hooks/useAuth';
import { AuthErrorProvider } from '@/components/layouts/LoginLayout/AuthError';
import { render } from '@/lib/test/render';

import { AuthButton } from '.';

it('should be rendered correctly', () => {
  const methods: authMethod[] = ['Google', 'Twitter', 'GitHub'];
  const types: authType[] = ['login', 'signup'];

  for (const method of methods) {
    for (const type of types) {
      const { container } = render(<AuthButton method={method} type={type} />, {
        wrapper: AuthErrorProvider,
      });
      expect(container.firstChild).toMatchSnapshot();
    }
  }
});

it('should have "button" role', () => {
  render(<AuthButton method="Google" type="login" />, {
    wrapper: AuthErrorProvider,
  });
  expect(screen.getByRole('button')).toBeInTheDocument();
});
