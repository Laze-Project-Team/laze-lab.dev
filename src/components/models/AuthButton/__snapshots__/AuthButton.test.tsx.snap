// Vitest Snapshot v1

exports[`should be rendered correctly 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`should be rendered correctly 2`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`should be rendered correctly 3`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`should be rendered correctly 4`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`should be rendered correctly 5`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`should be rendered correctly 6`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-lgsd8a-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  authwith
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
